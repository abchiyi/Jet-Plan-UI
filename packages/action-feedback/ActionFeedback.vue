<script>
function propInit(_type, _default) {
  return {
    type: _type || Boolean,
    default: _default || false
  };
}
import ripple from "./ActionFeedbackRipple.vue";
import { h, TransitionGroup } from "vue";
export default {
  name: "m-action-feedback",
  props: {
    // TODO 验证色彩字符串是否合法
    color: propInit(String, "#b3b3b3"),
    // TODO 或者透明度曲线
    opacity: propInit(String, "0.5"),
    tag: propInit(String, "div"),
    active: propInit(),
    hover: propInit(),
    ripple: propInit()
  },
  data() {
    return {
      data_hover: false,
      data_touch: false,
      masks: [],
      key: 0
    };
  },
  computed: {
    classes() {
      return [
        "m-action-feedback",
        this.hover && this.data_hover ? "hover" : ""
      ];
    }
  },
  methods: {
    // Ripple
    createRippleAttrs(event) {
      const ripple = {
        data: {
          opacity: this.opacity,
          el: this.$refs.self,
          color: this.color,
          event: event,
          ripple: !this.active
        },
        key: this.key++
      };
      return ripple;
    },
    removeRipple() {
      this.masks.forEach((ripple, index, masks) => {
        if (ripple.key != this.key) {
          masks.splice(index, 1);
        }
      });
    },
    // Render
    renderDefault() {
      const solt = this.$slots.default;
      if (solt) {
        return solt();
      }
      return "Submit";
    },
    renderRipples() {
      let key = 0;
      return h(
        TransitionGroup,
        { name: "mask" },
        {
          default: () =>
            this.masks.map(attrs => {
              return h(ripple, {
                data: attrs.data,
                key: key++
              });
            })
        }
      );
    },
    render() {
      return [this.renderDefault(), this.renderRipples()];
    },
    // Hover
    enter() {
      if (this.hover) this.data_hover = true;
    },
    leave() {
      if (this.hover) this.data_hover = false;
      this.removeRipple();
    },
    // Click
    startClick(event) {
      //   this.data_active = true;
      // Ripple
      if (this.ripple || this.active) {
        if (event.button === 0 && !this.ignoreClick) {
          this.masks.push(this.createRippleAttrs(event));
        }
        if (this.ignoreClick) this.ignoreClick = false;
      }
    },
    endClick() {
      if (this.data_touch) this.leave();
      //   this.data_active = false;
      this.removeRipple();
    },
    // Touch
    startTouche(event) {
      // XXX data_touch 不能设置为 false 未知的影响
      this.data_touch = true;
      this.enter();

      // Ripple
      if ((event.touches && this.ripple) || this.active) {
        this.masks.push(this.createRippleAttrs(event.touches[0]));
        // 触发"touche"事件时会在之后触发"click"事件
        // 此变量改变下一次"click"回调函数的运行结果
        this.ignoreClick = true;
      }
    },
    endTouche() {
      this.leave();
      this.removeRipple();
    }
  },
  watch: {},
  render() {
    return h(
      this.tag,
      {
        class: this.classes,
        onmousedown: this.startClick,
        onmouseup: this.endClick,

        onmouseenter: this.enter,
        onmouseleave: this.leave,

        ontouchstart: this.startTouche,
        ontouchcancel: this.endTouche,
        ontouchend: this.endTouche,
        ref: "self"
      },
      {
        default: this.render
      }
    );
  }
};
</script>
<style >
.m-action-feedback {
  position: relative;
  overflow: hidden;
}

.m-action-feedback::before {
  transition: 0.3s var(--ease-out-slow);
  background-color: var(--mask);
  position: absolute;
  content: "";
  opacity: 0;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
}

.m-action-feedback.hover:hover::before {
  opacity: 0.5;
}
</style>
