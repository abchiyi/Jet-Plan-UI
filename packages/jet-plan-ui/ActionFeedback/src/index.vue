<script lang="ts">
import baseAction from "./baseAction.vue";
import { h, defineComponent, Transition } from "vue";

const Mask = defineComponent({
  props: {
    styleConfig: {
      type: Object,
      require: true,
    },
  },
  render() {
    return h("span", {
      class: ["j-mask", "mask"],
      style: this.styleConfig,
    });
  },
});

export default defineComponent({
  name: "j-action-feedback",
  props: {
    name: { type: String, default: "j" }, //TODO 未添加
    hover: { type: Boolean, default: false },
    active: { type: Boolean, default: false },
    activeColor: {
      type: String,
      default: "var(--mask)",
    },
    activeOpacity: {
      type: Number,
      default: 1,
    },
  },
  data() {
    return {
      activated: false,
    };
  },
  methods: {
    handleAction(trigger: boolean) {
      if (!this.active) return;
      this.activated = trigger;
    },

    genMaskConfig() {
      return {
        background: this.activeColor,
        "--opacity": this.activeOpacity,
      };
    },
  },
  render() {
    const styleConfig = this.genMaskConfig();
    const Active = this.activated;
    const MaskActive = (() => {
      return h(
        Transition,
        { name: "mask" },
        {
          default() {
            if (!Active) return;
            return h(Mask, { styleConfig });
          },
        }
      );
    })();

    return h(
      baseAction,
      {
        class: [
          "j-action-feedback",
          this.hover ? "hover" : "",
          this.activated ? "active" : "",
        ],
        onActive_from: () => {
          this.handleAction(true);
        },
        onActive_to: () => {
          this.handleAction(false);
        },
      },
      {
        default: () => [this.$slots.default?.(), MaskActive],
      }
    );
  },
});
</script>
<style>
.j-action-feedback {
  outline: solid 3px #00000000;
  position: relative;
  transition: unset;
  overflow: hidden;
}

.j-action-feedback,
.j-action-feedback::after {
  transition: 0.3s var(--ease-out-slow);
}

.j-action-feedback::after {
  pointer-events: none;
  position: absolute;
  content: "";
  opacity: 0;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
}

/* Mouse hover */
.j-action-feedback.hover:hover::after {
  background-color: var(--mask);
}

.j-action-feedback.hover::after {
  opacity: 0.5;
}

@supports (-webkit-tap-highlight-color: #ffffff00) {
  .j-action-feedback {
    -webkit-tap-highlight-color: #ffffff00;
  }
}

/* Mask */
.j-action-feedback .j-mask {
  transform: translate3d(0, 0, 0);
  opacity: var(--opacity);
  pointer-events: none;
  position: absolute;
  user-select: none;
  display: block;
  height: 100%;
  width: 100%;
  left: 0;
  top: 0;
}

.j-action-feedback .mask-leave-to {
  transition: 0.4s cubic-bezier(0.6, 0.6, 0.4, 0.7);
  opacity: 0;
}
</style>
