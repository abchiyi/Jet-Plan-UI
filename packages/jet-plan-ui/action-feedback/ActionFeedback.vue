<script>
// TODO 修正 'TAB' 按键 选中时的样式应用问题
import { propInit, propInitBoolean } from '../tool/lib'
import baseAction from './baseAction.vue'
import mask from './ActionFeedbackMask.vue'
import { h, TransitionGroup } from 'vue'
export default {
    name: 'j-action-feedback',
    props: {
        name: propInit(String, 'j'),
        hover: propInitBoolean(false),
        active: propInitBoolean(false),
        hoverDelay: propInit(Number, 0),
    },
    data () {
        return {
            data_active: false,
            data_hover: false,
            masks: [],
            key: 0,
        }
    },
    computed: {
        classes () {
            return [
                `${this.name}-action-feedback`,
                this.data_hover && this.hover ? 'hover' : '',
                this.data_active && this.active ? 'active' : '',
            ]
        },
        // Render
        renderDefault () {
            const slot = this.$slots.default
            if (slot) {
                return slot()
            }
            return 'Submit'
        },
        renderMask () {
            let key = 0
            return h(
                TransitionGroup,
                { name: 'mask' },
                {
                    default: () =>
                        this.masks.map((attrs) => {
                            return h(mask, {
                                data: attrs.data,
                                key: key++,
                            })
                        }),
                }
            )
        },
    },
    methods: {
        createMask (event) {
            return {
                data: {
                    opacity: 1,
                    el: this.$refs.self,
                    color: 'var(--mask)',
                    event: event,
                },
                key: this.key++,
            }
        },
        removeMask () {
            this.masks = []
        },
        handlerActive (event) {
            if (event.active) {
                // 遮罩在自定义类名中不启用
                if (this.name == 'j') {
                    let e = event.event
                    if (this.active && (e.button === 0 || e.touches)) {
                        this.masks.push(this.createMask(e))
                    }
                }
            } else {
                this.removeMask()
            }

            this.data_active = event.active
        },
        handlerHover (event) {
            if (this.hover) {
                if (event.active) {
                    this.$hover_timeout_id = setTimeout(() => {
                        this.data_hover = true
                    }, this.hoverDelay)
                } else {
                    clearTimeout(this.$hover_timeout_id)
                    this.data_hover = false
                }
            }
        },
    },
    render () {
        return h(
            baseAction,
            {
                class: [...this.classes],

                onActive_from: this.handlerActive,
                onActive_to: this.handlerActive,

                onHover_from: this.handlerHover,
                onHover_to: this.handlerHover,
            },
            {
                default: () => [this.renderDefault, this.renderMask],
            }
        )
    },
}
</script>
<style>
.j-action-feedback {
    position: relative;
    outline: unset;
}

.j-action-feedback::after,
.j-action-feedback .j-mask {
    border-radius: inherit;
}

.j-action-feedback::after {
    transition: 0.2s var(--ease-out-slow);
}

.j-action-feedback::after {
    pointer-events: none;
    position: absolute;
    content: '';
    bottom: 0;
    right: 0;
    left: 0;
    top: 0;
}

/* Mouse hover */
.j-action-feedback.hover::after {
    background-color: var(--mask);
}
</style>
