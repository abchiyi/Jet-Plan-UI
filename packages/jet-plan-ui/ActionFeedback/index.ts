import type { App, Plugin } from "vue";
import ActionFeedback from "./src/index.vue";

export const ActionFeedbackPlugin: Plugin = {
  install(app: App) {
    app.component("j-action-feedback", ActionFeedback);
  },
};

export { ActionFeedback };
