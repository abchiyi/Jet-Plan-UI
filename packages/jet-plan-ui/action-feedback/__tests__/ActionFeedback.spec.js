import ActionFeedback from '../ActionFeedback.vue'

import {
    mount,
    // shallowMount
} from '@vue/test-utils'

function cssAttrsStringToObj(str) {
    const temp = {}
    const attrs = str.split(';');
    for (let n = 0; n < attrs.length; n++) {
        if (attrs[n].trim()) {
            let [key, value] = attrs[n].split(':')
            temp[key.trim()] = value.trim();
        }
    }
    return temp
}

describe('ActionFeedback.vue', () => {
    it('Style test', async () => {
        const wrapper = mount(ActionFeedback, {
            propsData: {
                active: true
            }
        })
        expect(wrapper.html()).toMatchSnapshot()

        await wrapper.setProps({
            active: true,
            hover: true,
        })
        expect(wrapper.html()).toMatchSnapshot()

        await wrapper.setProps({
            active: true,
            hover: true,
            focus: true
        })

        expect(wrapper.html()).toMatchSnapshot()

        await wrapper.setProps({
            active: true,
            hover: true,
            focus: true,
            focusOutline: true,
        })
        expect(wrapper.html()).toMatchSnapshot()

        await wrapper.setProps({
            colorFocusOutline: '#fff',
            colorActive: '#263238',
            colorFocus: '#5a7fa2',
            colorHover: "#a3b3d4"
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it("Custom color", async () => {

        const wrapper = mount(ActionFeedback)

        const colorData = {
            colorFocusOutline: '#fff',
            colorActive: '#263238',
            colorFocus: '#5a7fa2',
            colorHover: "#a3b3d4"
        }

        await wrapper.setProps(colorData)

        const style = cssAttrsStringToObj(
            wrapper.attributes('style')
        )
        expect(style['--color-focus-out-line']).toEqual(
            colorData.colorFocusOutline
        )
        expect(style['--color-focus']).toEqual(
            colorData.colorFocus
        )
        expect(style['--color-hover']).toEqual(
            colorData.colorHover
        )
    })

    it('mask opacity', async () => {
        const wrapper = mount(ActionFeedback)
        const style = cssAttrsStringToObj(
            wrapper.attributes('style')
        )
        expect(style['--mask-opacity']).toEqual('0.5')

        await wrapper.setProps({
            opacity: 0.2
        })
        expect(cssAttrsStringToObj(
            wrapper.attributes('style')
        )['--mask-opacity']).toEqual('0.2')

    })

    it('Custom class', async () => {
        const wrapper = mount(ActionFeedback)

        expect(wrapper.vm.name).toEqual('j')



    })
})
