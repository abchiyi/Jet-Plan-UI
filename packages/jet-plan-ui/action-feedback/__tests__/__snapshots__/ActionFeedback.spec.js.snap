// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionFeedback.vue Style test 1`] = `
<div class="j-base-action j-action-feedback" ontouchstart="function () { [native code] }" ontouchcancel="function () { [native code] }" ontouchend="function () { [native code] }" style="--color-focus: var(--mask); --color-hover: var(--mask); --mask-opacity: 0.5;">Submit<transition-group-stub></transition-group-stub>
</div>
`;

exports[`ActionFeedback.vue Style test 2`] = `
<div class="j-base-action j-action-feedback" ontouchstart="function () { [native code] }" ontouchcancel="function () { [native code] }" ontouchend="function () { [native code] }" style="--color-focus: var(--mask); --color-hover: var(--mask); --mask-opacity: 0.5;">Submit<transition-group-stub></transition-group-stub>
</div>
`;

exports[`ActionFeedback.vue Style test 3`] = `
<div class="j-base-action j-action-feedback" ontouchstart="function () { [native code] }" ontouchcancel="function () { [native code] }" ontouchend="function () { [native code] }" style="--color-focus: var(--mask); --color-hover: var(--mask); --mask-opacity: 0.5;">Submit<transition-group-stub></transition-group-stub>
</div>
`;

exports[`ActionFeedback.vue Style test 4`] = `
<div class="j-base-action j-action-feedback" ontouchstart="function () { [native code] }" ontouchcancel="function () { [native code] }" ontouchend="function () { [native code] }" style="--color-focus: var(--mask); --color-hover: var(--mask); --mask-opacity: 0.5;">Submit<transition-group-stub></transition-group-stub>
</div>
`;

exports[`ActionFeedback.vue Style test 5`] = `
<div class="j-base-action j-action-feedback" ontouchstart="function () { [native code] }" ontouchcancel="function () { [native code] }" ontouchend="function () { [native code] }" style="--color-focus: #5a7fa2; --color-hover: #a3b3d4; --mask-opacity: 0.5;">Submit<transition-group-stub></transition-group-stub>
</div>
`;
